### DISCUSSION TITLE
为什么我们的工具栏参数需要加一个 `--`？

### DISCUSSION THREAD

**Discussion by mark9804** [Arondight/Adachi-BOT#963] [Q&A] [answered]
比如说

`npm run tool-resources -- -h`

感觉这两道短横不需要？

```ts
const args = process.argv.slice(2);

console.log(args);
```

```sh
$ pnpm start -n

> razor@1.0.0 start /Users/mark_chen/Documents/GitHub.nosync/razor
> tsc --build && node dist/index.js "-n"

[ '-n' ]
```

这样子也可以直接读取参数

**Comment by Arondight** [comment, marked as answer]
## 文档

From `man 1 getopt` ...

```
The special argument  "--" forces an end of option-scanning regardless of the scanning mode.
```

## 人话

约定当处理参数的库遇到选项 `--` 时，表示其后的所有参数均不是选项。

## 实战

```shell
pushd /tmp/
touch -- -h
ls -l -- -h
rm -vf -- -h
popd
```

**Reply by Arondight** [reply]
## 补充

1. `npm run tool-resources -- -h` : npm 命令没有选项，最后的 `-h` 是一个普通的字符串参数（最后传递给了 `./tool/resources.js`）
2. `npm run tool-resources -h` : npm 发现 `-h` 选项，打印了 npm 命令的对应帮助信息

**Reply by mark9804** [reply]
就是后面的选项由我们的脚本处理而不是由 npm 处理？如果这样的话大概理解了

**Reply by Arondight** [reply]
> 就是后面的选项由我们的脚本处理而不是由 npm 处理？如果这样的话大概理解了

说的直白一点，每个参数处理的库（例如 getopt）都会有一个选项和逻辑之间的映射，所有在 `--` 后的参数均不匹配这个映射，至于程序怎么处理是程序员自己的事，反正库不处理（这里程序把其后的字符串都用做了目标脚本的参数）

**Reply by mark9804** [reply]
拿 minimist 库试了一下，应该是理解了，谢谢老大哥
